src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/CreateDream.kt

```diff
@@ -28,12 +37,29 @@ class CreateDream(
       .flatMap { dream ->
         findUserById.execute(userId)
           .toEither { DreamCreationErrors.CreatorNotFound(userId) }
-          .map { user ->
-            dreamRequest.tags.forEach { tag ->
-              createTag.execute(tag).map { dream.tags.add(it) }
-            }
-            dao.save(dream.toEntity(user))
+          .flatMap { user ->
+            dreamRequest.tags
+              .map(createTag::execute).k()
+              .sequence(Either.applicative()).fix()
```

Comment:
You may be wondering what is this all about.

Well, executing `dreamRequest.tags.map(createTag::execute)` returns a `List<Either<TagCreationError, Tag>>` and we need a list of tags for adding them to the dream we are saving in database.

We have two options here. We could iterate over the list and handle each Either separately which can be a pain. Or we can use some functional big words to help us ease this problem 😎.

Firstly we need to know that `List` has an instance for the `Traverse` typeclass. This means list can be traversed over from left to right applying a function on each element. If we apply the identity function (which is what `sequence` is doing) we can turn around the type so it transforms into `Either<TagCreationError, List<Tag>>` and we can work with this now.

Reply:
I have a few doubts @Cotel:
* With `.k()` you're converting `List<Either<TagCreationError, Tag>>` into `ListK<Either<TagCreationError, Tag>>` ?
* You say _"...we can turn around the type so it transforms into..."_ this is done with the `Either.applicative()` or what this does?
* What does `.fix()` ?

Reply:
`.fix()` parse a `Kind<EitherOf, Error, Success>` to `Either<Error, Success` 👍 it is like parsing from internal representation of abstraction to the concrete type 👍

Reply:
`ListK` is a wrapper around kotlin stdlib`List` which has more functionalities related with typeclasses and so. We need it because of the `Traverse` operations for lists.

We use `Either.applicative()` because an `Applicative` is a typeclass which can wrap functions in containers (let's call it that way for simplifying). If we look at the `traverse` function of `Traversable` it has this shape:

```kotlin
fun <G, A, B> Kind<F, A>.traverse(AP: Applicative<G>, f: (A) -> Kind<G, B>): Kind<G, Kind<F, B>>
```

`traverse` is a lot like `map` but it let us run effects. We need an Applicative instance for accumulating those effects while. For example:

```
fun half(x: Int) = x / 2
val list = [2, 4, 6, 8]
list.map(::half)
// [1, 2, 3, 4]
``` 

This is 100% pure, but what if we want to make this more safe? We can wrap the `half` results in a `Option` datatype if the number is not even.

```
fun half(x: Int) = if (x % 2 == 0) Option.just(x/2) else Option.empty()
val list = [2, 4, 6, 8]
list.map(::half)
// [Just(1), Just(2), Just(3), Just(4)]

val list2 = [1, 2, 3, 4]
list2.map(::half)
// [Empty, Just(1), Empty, Just(2)]
``` 

This is 100% pure again. But probably this structure is not handy for us to work with it. So we can use `traverse` for transform it into `Option<List<Int>>` which is better. But, what if any of the `Options` inside the list is `Empty`? Merging all the options in just one can be considered as an effect because we are unwrapping them. So we need the Applicative for shortcircuiting (is that a word?).

```kotlin
list.traverse(::half)
// Just([1, 2, 3, 4])

list2.traverse(::half)
// Empty
```

Hope it is clear 😅. You can find more information in Haskell [here](https://wiki.haskell.org/Typeclassopedia), or in Scala [here](https://gist.github.com/jdegoes/97459c0045f373f4eaf126998d8f65dc) and hopefully Arrow's docs about Traverse will be finished soon [here](https://arrow-kt.io/docs/typeclasses/traverse/)

Reply:
Thanks @Cotel